src/coreclr/vm/methodtablebuilder.cpp

```diff
@@ -8677,30 +8677,28 @@ MethodTableBuilder::HandleExplicitLayout(
             else
             {
                 MethodTable *pByValueMT = pByValueClassCache[valueClassCacheIndex];
-                if (pByValueMT->IsByRefLike() || pByValueMT->ContainsGCPointers())
+                BOOL hasGcPointers = pByValueMT->ContainsGCPointers();
+                if (hasGcPointers || pByValueMT->IsByRefLike())
                 {
-                    if ((pFD->GetOffset() & ((ULONG)TARGET_POINTER_SIZE - 1)) != 0)
+                    if (hasGcPointers && ((pFD->GetOffset() & ((ULONG)TARGET_POINTER_SIZE - 1)) != 0))
```

Comment:
This whole if-statement should not be needed. The same error condition should be checked by `CheckValueClassLayout` too - just a bit slower - but it does not matter since we are going to fail anyway.

Reply:
Sorry, I got confused between `CheckValueClassLayout` and `CheckByRefLikeValueClassLayout`. 

For byref-like types, I think this check is redundant with https://github.com/dotnet/runtime/pull/111584/files#diff-4b65f34bf68fdb80cc734cc2435c72ef7f86ef16b9f56af47c06b87761383d3aR8969-R8972 . I agree that this check is needed for types that are contain GC references, but that are not byref-like types.

It may be less confusing to move this alignment check to https://github.com/dotnet/runtime/pull/111584/files#diff-4b65f34bf68fdb80cc734cc2435c72ef7f86ef16b9f56af47c06b87761383d3aR8877 so that CheckValueClassLayout is responsible for the alignment check for both byref-like and non-byref-like types?

Reply:
Excellent suggestions. Please take a look at latest. I hope I was able to capture this suggestion.